### PR TITLE
Display performance breakdown in a tooltip

### LINKED ISSUE #15570: Display a breakdown of the user's PP in the results screen
### Discussed in https://github.com/ppy/osu/discussions/15566

<div type='discussions-op-text'>

<sup>Originally posted by **developomp** November 11, 2021</sup>
Show breakdown of pp (aim, speed, accuracy, etc.) in the result screen.
</div>

### PR DESCRIPTION
**Pull request by hlysine**
Closes #15570 

~Depends on #16575 for correct breakdown values~ No longer the case

~Add a tooltip to `PerformanceStatistic` displaying pp breakdown.~ See updates below

<details>
<summary>Old screenshot</summary>

osu! 17_1_2022 9_26_01 pm (2)

</details>

# Updates

Here's a summary of major changes since the initial commit so reviewers/newcomers don't have to scroll through the entire conversation:

- Instead of showing the absolute attribute values, a perfect play is created based on the beatmap, ruleset, and mods selected, then the achieved performance is compared with the perfect play performance to generate a percentage for each attribute.
- Currently, this comparison assumes that attribute values are combined linearly to generate the final PP value, which is a close approximation, but not exactly the case. However, correcting for this non-linear combination requires refactoring the performance calculator, which is not included in this PR.
- The entire chart has been moved from a tooltip to the expanded results screen for better visibility. The expanded results screen has been refactored to support items like this one that don't require watching a replay to display.
- The performance breakdown revealed that osu diffcalc doesn't properly support lazer-first scores (at least in terms of max combo calculation). A local function has been added in `PerformanceBreakdownCalculator` to circumvent this. The fix to osu diffcalc is tracked in #16575.

<details>
<summary>Screenshots</summary>

image

image

</details>

### CODE REVIEW

**Comment by mk56-spn** [comment]
holy you are fast, looks rather primitive but def works

**Comment by mk56-spn** [comment]
> The numbers make no sense though? The way this tooltip is structured suggests that the values next to the bars should total to the sum. But they don't, even on the screenshot above 71+36+90 sums up to 197 and not 202. And that's because the constituent parts are combined in a non-linear manner to the final pp.
> 
> Like the raw performance attribute values don't matter to anyone besides maybe pp developers, because they're _not actual pp points_ - so why show them?
> 
> The way the bars are linearly scaled to the sum of the attributes is also questionable. There's no standardisation in place that says that the attributes must be even anywhere near the same magnitude.

because people have  been wanting to see the constituent parts of PP forever, to tell at a glance what a plays strongest areas are, simple as

**Comment by bdach** [comment]
...except for the part where this isn't the actual breakdown of a performance? precisely because the attributes are combined into the final value in a non-trivial manner? the bar graph is actively misleading. it makes you think the final pp value is the sum of the breakdown parts. while it may be nowhere near.

here's a taiko example:

osu_2022-01-17_15-45-22

and a particularly egregious mania example:

osu_2022-01-17_15-42-18

the fact that a feature is oft-requested does not mean that it has to be implemented by all costs when it's not really functioning as expected.

**Comment by frenzibyte** [comment]
What would make sense for having these is when comparing them against a perfect play, as already mentioned in https://github.com/ppy/osu/discussions/15566#discussioncomment-1623891 instead of what seems to be just an arbitrary width to the bar, and also not consider the final PP value as the direct "Total" as mentioned above.

**Comment by bdach** [comment]
i agree the above sounds workable. i'd imagine that would work in the following way:

take the score's performance attributes, compute performance attributes for a perfect play, and then for each performance attribute fix other attributes to the perfect value and then calculate the pp for that score. the amount of pp lost on this skill relative to the perfect play would work in a way for a breakdown.

even then i probably wouldn't show raw numbers because they're likely to also make no sense. maybe each attribute could be ranked by percentage or something, or have a letter rank attached.

**Comment by peppy** [comment]
Percentage is what I would have proposed.

**Comment by Henry-YSLin** [comment]
The bar chart is now changed to compare performance attributes with a perfect play:

osu! 18_1_2022 10_13_35 pm

To avoid misconceptions, this breakdown only shows you how a score could be improved when compared to a perfect play of the same map. An SS play would simply have 100% across the board because it _is_ a perfect play. Visualizing star rating attributes (#9463) should give you a better understanding of the skills demonstrated in a play.

File: osu.Game/Screens/Ranking/Expanded/Statistics/PerformanceStatistic.cs
Comment (bdach) [inline osu.Game/Screens/Ranking/Expanded/Statistics/PerformanceStatistic.cs:59]: not super sure all of this logic should reside in a drawable such as this, especially so that the data is just shipped down to the tooltip. would probably consider splitting off a separate component to compute this - could end up needing to reuse it later in other contexts, too.

File: osu.Game/Screens/Ranking/Expanded/Statistics/PerformanceStatistic.cs
Comment (bdach) [inline osu.Game/Screens/Ranking/Expanded/Statistics/PerformanceStatistic.cs:91]: falling back to the score's max combo looks questionable - there's no actual guarantee it is the actual max combo of the map, is there? if the map isn't locally available to compute the combo, then the tooltip shouldn't be displayed at all.

File: osu.Game/Screens/Ranking/Expanded/Statistics/PerformanceStatisticTooltip.cs
Comment (bdach) [inline osu.Game/Screens/Ranking/Expanded/Statistics/PerformanceStatisticTooltip.cs:87]: this is better, but isn't quite what i had in mind [in my earlier comment](https://github.com/ppy/osu/pull/16483#issuecomment-1014627537). while displaying the user's score performance attributes as a ratio relative to the perfect score's, there's still sort of an issue in that there is no actual guarantee that the pp algorithms are going to give equal weighting of the attributes, or that the final pp value is going to be a linear combination of the attribute values.

to substantiate this: here's how 3/4 "official" rulesets compute the final pp currently.

https://github.com/ppy/osu/blob/b3983db69d5a5646cae2c3bdd9db326ea472fbfd/osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs#L66-L72
https://github.com/ppy/osu/blob/b3983db69d5a5646cae2c3bdd9db326ea472fbfd/osu.Game.Rulesets.Taiko/Difficulty/TaikoPerformanceCalculator.cs#L48-L52
https://github.com/ppy/osu/blob/b3983db69d5a5646cae2c3bdd9db326ea472fbfd/osu.Game.Rulesets.Mania/Difficulty/ManiaPerformanceCalculator.cs#L66-L70

notice that these all use the [generalized / power mean](https://en.wikipedia.org/wiki/Generalized_mean) of the attribute values to compute the final value. and the thing with the power mean is that it's non linear.

below is a graph; red illustrates the power mean, and blue illustrates the linear sum. the interpretation of this graph is: "each point on the red/blue curve is such that if the x and y coordinates are plugged into the correct curve's formula, the result is 10". in other words, moving that to the pp domain - that's what the values of performance attributes would have to be for the final pp value to be 10 according to each curve's formula.

2022-01-18-223159_965x430_scrot

note that the red curve is not linear flat, it's slightly convex. that's the nonlinearity.

to get to the point: that's why i proposed the original scheme, to eliminate the impact of the choice of combining function on how the breakdown works. because imagine that you may want to raise an attribute to another attribute's power, or similar - that isn't going to translate well to the current implementation, because seemingly small variations of a performance attribute may wildly influence the final pp value. so what i propose instead is, formally:

let f(x<sub>1</sub>, x<sub>2</sub>, ..., x<sub>n</sub>) be the function used to combine values of performance attributes x<sub>1</sub>, x<sub>2</sub>, ..., x<sub>n</sub> into a final pp value; in our current situation, it would be the power mean. now let p<sub>1</sub>, p<sub>2</sub>, ..., p<sub>n</sub> be the value of performance attributes for a perfect play on the same map. i propose that the breakdown's value for the i-th attribute should equal:

b<sub>i</sub> = f(p<sub>1</sub>, p<sub>2</sub>, ..., p<sub>i - 1</sub>, x<sub>i</sub>, p<sub>i + 1</sub> ..., p<sub>n</sub>) / f(p<sub>1</sub>, p<sub>2</sub>, ..., p<sub>i - 1</sub>, p<sub>i</sub>, p<sub>i + 1</sub> ..., p<sub>n</sub>)

this would *roughly* represent the number of pp that the player lost on the i-th attribute alone if they played perfectly with respect to every other attribute, compared to if they were playing 100% perfectly in general.

i think this is the best that can be done with respect to separating out attributes' influence on the final pp value, without constraining how the combining function f works and looks.

Comment (bdach) [reply osu.Game/Screens/Ranking/Expanded/Statistics/PerformanceStatisticTooltip.cs:87]: had some discussion about this elsewhere and it was rightly pointed out to me that the above can have a weird artifact wherein fixing the i-th attribute to 0 might not yield 0 with the proposed scheme. this could be remedied by having that as the absolute minimum (as in, the evaluation scale goes from "perfect score with i-th attribute fixed to 0" to "perfect score"), i.e. in the formula above, you'd subtract f(p<sub>1</sub>, p<sub>2</sub>, ..., p<sub>i - 1</sub>, 0, p<sub>i + 1</sub> ..., p<sub>n</sub>) from both the numerator and the denominator.

Comment (Henry-YSLin) [reply osu.Game/Screens/Ranking/Expanded/Statistics/PerformanceStatisticTooltip.cs:85]: I do agree that this is a better scheme than simple ratios, but the performance calculator needs to be refactored for this scheme to work. I'd rather do this in a separate PR after this one's merged.

**Comment by Moltanicaa** [comment]
Will this work for all rulesets?

**Comment by Henry-YSLin** [comment]
Yes, except catch which doesn't have separate attributes in the performance calculator

File: osu.Game/Rulesets/Difficulty/PerformanceBreakdownCalculator.cs
Comment (bdach) [inline osu.Game/Rulesets/Difficulty/PerformanceBreakdownCalculator.cs:16]: Could make this a `Component`, at which point you could  use DI the same way as you would from a standard drawable, so long as the calculator is `Add()`ed to another composite drawable. Maybe not necesssary though.

File: osu.Game/Rulesets/Difficulty/PerformanceBreakdownCalculator.cs
Comment (bdach) [inline osu.Game/Rulesets/Difficulty/PerformanceBreakdownCalculator.cs:59]: Why is this using a blocking call rather than awaits? Also how does `GetResultSafely()` not crash here? `Task.Factory.StartNew()` should queue up the work inside onto a TPL thread, and this `GetResultSafely()` will block on an operation that's potentially taking place on another pool thread, possibly triggering a deadlock? I'd understand if the `Task.Factory.StartNew()` call specified `LongRunning` but it doesn't...?

The only explanation I have right now is that the result always already happens to be available because of prior computations (on song select or otherwise). Either way I'd say this should use a proper async flow.

File: osu.Game/Scoring/ScorePerformanceCache.cs
Comment (bdach) [inline osu.Game/Scoring/ScorePerformanceCache.cs:19]: Not sure about this change. This way every score PP calculation will store a copy of the "perfect play" attribute values. So there'll be N copies of the same data for N user scores, and each copy will be calculated from scratch. Seems kinda wasteful.

File: osu.Game/Screens/Ranking/Expanded/Statistics/PerformanceStatisticTooltip.cs
Comment (bdach) [inline osu.Game/Screens/Ranking/Expanded/Statistics/PerformanceStatisticTooltip.cs:62]: This conditional should really live in `SetContent()`. `PopIn()` shouldn't contain actual logic if it can be avoided, just the transition definition.

File: osu.Game/Scoring/ScorePerformanceCache.cs
Comment (Henry-YSLin) [reply osu.Game/Scoring/ScorePerformanceCache.cs:19]: Would you prefer a separate cache for perfect play performance or computing it on the fly without caching? I don't think a cache is necessary but I could be wrong seeing that `ScorePerformanceCache` is created with `PerformanceStatistic` being its only usage.

File: osu.Game/Rulesets/Difficulty/PerformanceBreakdownCalculator.cs
Comment (Henry-YSLin) [reply osu.Game/Rulesets/Difficulty/PerformanceBreakdownCalculator.cs:16]: Probably not necessary. Feels weird to add a calculator to draw hierarchy just for DI.

File: osu.Game/Scoring/ScorePerformanceCache.cs
Comment (bdach) [reply osu.Game/Scoring/ScorePerformanceCache.cs:19]: Caching would be optimal (because you'd only have to cache one per difficulty) but upon checking it seems to be not-horrible performance wise even on marathon maps like https://osu.ppy.sh/beatmapsets/29157#osu/156352 so maybe fine to keep this as is for now.

File: osu.Game/Screens/Ranking/Expanded/Statistics/PerformanceStatistic.cs
Comment (bdach) [inline osu.Game/Screens/Ranking/Expanded/Statistics/PerformanceStatistic.cs:49]: The schedule on `setPerformanceValue()` that was there on `master` was removed, causing the following error to potentially show up:

```
[runtime] 2022-01-20 21:03:47 [error]: System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. (Cannot mutate the Transforms of a Loaded Drawable while not on the update thread. Consider using Schedule to schedule the mutation operation.)
[runtime] 2022-01-20 21:03:47 [error]: ---> osu.Framework.Graphics.Drawable+InvalidThreadForMutationException: Cannot mutate the Transforms of a Loaded Drawable while not on the update thread. Consider using Schedule to schedule the mutation operation.
[runtime] 2022-01-20 21:03:47 [error]: at osu.Framework.Graphics.Drawable.EnsureMutationAllowed(String member)
[runtime] 2022-01-20 21:03:47 [error]: at osu.Framework.Graphics.Transforms.Transformable.AddTransform(Transform transform, Nullable`1 customTransformID)
[runtime] 2022-01-20 21:03:47 [error]: at osu.Framework.Graphics.TransformableExtensions.TransformTo[TThis](TThis t, Transform transform)
[runtime] 2022-01-20 21:03:47 [error]: at osu.Framework.Graphics.TransformableExtensions.TransformTo[TThis,TValue](TThis t, String propertyOrFieldName, TValue newValue, Double duration, Easing easing)
[runtime] 2022-01-20 21:03:47 [error]: at osu.Game.Graphics.UserInterface.RollingCounter`1.TransformCount(T currentValue, T newValue) in /home/dachb/Dokumenty/opensource/osu/osu.Game/Graphics/UserInterface/RollingCounter.cs:line 162
[runtime] 2022-01-20 21:03:47 [error]: at osu.Game.Graphics.UserInterface.RollingCounter`1.<LoadComplete>b__22_0(ValueChangedEvent`1 val) in /home/dachb/Dokumenty/opensource/osu/osu.Game/Graphics/UserInterface/RollingCounter.cs:line 93
[runtime] 2022-01-20 21:03:47 [error]: at osu.Framework.Bindables.Bindable`1.TriggerValueChange(T previousValue, Bindable`1 source, Boolean propagateToBindings, Boolean bypassChecks)
[runtime] 2022-01-20 21:03:47 [error]: at osu.Framework.Bindables.Bindable`1.TriggerValueChange(T previousValue, Bindable`1 source, Boolean propagateToBindings, Boolean bypassChecks)
[runtime] 2022-01-20 21:03:47 [error]: at osu.Framework.Bindables.Bindable`1.set_Value(T value)
[runtime] 2022-01-20 21:03:47 [error]: at osu.Game.Screens.Ranking.Expanded.Statistics.PerformanceStatistic.setPerformanceValue(PerformanceBreakdown breakdown) in /home/dachb/Dokumenty/opensource/osu/osu.Game/Screens/Ranking/Expanded/Statistics/PerformanceStatistic.cs:line 58
[runtime] 2022-01-20 21:03:47 [error]: at osu.Game.Screens.Ranking.Expanded.Statistics.PerformanceStatistic.<load>b__17_0(Task`1 t) in /home/dachb/Dokumenty/opensource/osu/osu.Game/Screens/Ranking/Expanded/Statistics/PerformanceStatistic.cs:line 49
```

Especially visible on maps such as https://osu.ppy.sh/beatmapsets/29157#osu/156352.

**Review by bdach — DISMISSED** [review]

File: osu.Game/Screens/Ranking/Expanded/Statistics/PerformanceStatistic.cs
Comment (Henry-YSLin) [reply osu.Game/Screens/Ranking/Expanded/Statistics/PerformanceStatistic.cs]: A careless mistake. Fixed now

**Comment by bdach** [comment]
I've tested this some more and I'm generally okay with this I think, but I would still like to get a second fresh opinion from someone who hasn't seen this diff this many times yet.

There's one breakage - in playlists, sometimes score panels are used even when the beatmap isn't locally available, but the current checks for that don't quite work and therefore the tooltip still shows with some rather nonsensical results:

osu_2022-01-22_14-43-53

but the underlying general issue there is #16480 and therefore fixing this here is probably out of scope...?

**Comment by Henry-YSLin** [comment]
Found a pretty serious problem with the performance breakdown calculation. The breakdown is completely inaccurate for a score set without classic mod in the osu ruleset. From the screenshot below, it is clear that the score should have perfect performance, but the performance breakdown just outputs bogus values.

image

I have a rough idea why this is happening, but I haven't got time to fix it right now. Gonna set this PR to draft for now to prevent premature merging.

**Comment by bdach** [comment]
It's probably due to combo scaling. Classic mod being enabled influences possible max combo.

**Comment by peppy** [comment]
Just going to mention that the "x/y" display next to final pp is hard to understand. I can't even imagine what it's intending to convey.

**Comment by smoogipoo** [comment]
> I can't even imagine what it's intending to convey.

The maximum PP. Perhaps you're confused because the screenshot was taken without the existence of the classic mod, and diffcalc doesn't currently support lazer-first scores (see: https://github.com/ppy/osu/pull/16575). It would be 198/198 for a perfect score.

**Comment by peppy** [comment]
Wow okay. Maybe I'm just stupid.

But I also wouldn't display it like that. The maximum should be its own row, or in parenthesis or something. As a fraction reads really badly IMHO.

**Comment by peppy** [comment]
I think it's okay. Someone may be able to come up with better terminology for the text.

**Comment by mk56-spn** [comment]
> Does this look better? I have to remove the bar in the middle, otherwise it looks odd being stuck between 2 lines.
> 
> image
> 

Imo achieved PP sounds better than final PP and juxtaposes itself better with maximum PP

**Comment by ikkue** [comment]
> Does this look better? I have to remove the bar in the middle, otherwise it looks odd being stuck between 2 lines.
> 
> image

I don't know if this is possible or a good idea, but I think plays without the Flashlight mod applied (which is like the majority of plays) should just have the bar and the percentage on the "Flashlight Bonus" row be replaced with the text "Flashlight not applied" or something similar.

**Comment by bdach** [comment]
> I don't know if this is possible or a good idea, but I think plays without the Flashlight mod applied (which is like the majority of plays) should just have the bar and the percentage on the "Flashlight Bonus" row be replaced with the text "Flashlight not applied" or something similar.

You could just not show a skill on the graph if the perfect play with the given beatmap and mods receives nothing pp-wise from that given skill. Assuming non-FL plays universally get 0 FL bonus this would have the desired effect.

**Comment by peppy** [comment]
Agree with hiding zero values.

**Comment by jai-x** [comment]
This may be an unwanted drive-by comment, but is there a reason this extra info is being implemented as a tooltip rather than on the expanded score and statistics display that shows when you click on a score card? It seems like a more logical place to put this kind of info?

**Comment by peppy** [comment]
Also agree that it is best moved to the expanded results screen. Displaying this kind of information in a tooltip is obtuse and hidden, and we added the expanded screen specifically for statistics like this.

**Comment by hlysine** [comment]
> Agree with hiding zero values.

Alright, will do.

> Also agree that it is best moved to the expanded results screen.

I reckon that would be the ideal place too, but currently the expanded results screen requires you to watch the replay before showing the stats, while pp breakdown doesn't really require the replay to calculate. Wouldn't it be too annoying to require users to do extra useless work before seeing the breakdown?

**Comment by peppy** [comment]
The requirement for watching replay can be moved locally to the components which require it.

**Comment by hlysine** [comment]
The entire chart has been moved to the expanded results screen:

<details>
<summary>Screenshots</summary>

image

image

</details>

Also, in an effort to unblock this PR, I added a function to calculate beatmap max combo locally in `PerformanceBreakdownCalculator`. It is essentially the same function as in #16575, which allows the effect of classic mod to be properly considered. As a result, the breakdown chart now displays results as expected (autoplay gets all 100%, no more above-100% weirdness).

Although the absolute PP value is still inaccurate (that would have to rely on #16575 to be fixed), the value on the chart matches those displayed in other places (e.g. performance statistic in the score panel), so it should be fine to go ahead with this PR for now. Gonna mark this as ready for review.

**Comment by peppy** [comment]
I think this seems fine for an initial effort. It's ugly, but so are the other statistics displays. I figure we'll do a design pass on all these in a grouped effort in the future.

File: osu.Game/Screens/Ranking/Statistics/PerformanceBreakdownChart.cs
Comment (peppy) [inline osu.Game/Screens/Ranking/Statistics/PerformanceBreakdownChart.cs:196]: Usually we'd check the divisor ahead of time rather than post-checking for NaN.

Out of curiosity, what's the case where this can happen? A beatmap with no objects somehow reaching results?

File: osu.Game/Screens/Ranking/Statistics/PerformanceBreakdownChart.cs
Comment (peppy) [inline osu.Game/Screens/Ranking/Statistics/PerformanceBreakdownChart.cs:150]: Feels like this should maybe be moved to `LoadComplete` to avoid potentially calculating before the user requests details statistics. But I guess it's fine as-is until we notice an issue.

File: osu.Game/Screens/Ranking/Statistics/PerformanceBreakdownChart.cs
Comment (peppy) [inline osu.Game/Screens/Ranking/Statistics/PerformanceBreakdownChart.cs:155]: I haven't managed to get this spinner to ever appear so it's hard to test whether it displays correctly or not. Were you manually adding delays to test this out?

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/Ranking/Statistics/PerformanceBreakdownChart.cs
Comment (hlysine) [reply osu.Game/Screens/Ranking/Statistics/PerformanceBreakdownChart.cs]: > Usually we'd check the divisor ahead of time rather than post-checking for NaN.

Alright, changed to a pre-check.

> Out of curiosity, what's the case where this can happen?

This is to hide attributes that have 0 as the maximum. e.g. if flashlight mod isn't on, flashlight bonus would always be 0.

File: osu.Game/Screens/Ranking/Statistics/PerformanceBreakdownChart.cs
Comment (hlysine) [reply osu.Game/Screens/Ranking/Statistics/PerformanceBreakdownChart.cs:151]: In my testing, it only starts calculation when the detailed statistics panel is first shown, so it should be fine.

File: osu.Game/Screens/Ranking/Statistics/PerformanceBreakdownChart.cs
Comment (hlysine) [reply osu.Game/Screens/Ranking/Statistics/PerformanceBreakdownChart.cs:156]: Yep. It is working fine. The chart is taller than the spinner, causing a height change when it finishes loading, but I guess it's not noticeable anyway.

**Review by peppy — APPROVED** [review]
Let's get this one in, I think it looks fine.

**Pull request merged**
